SSE listen

wtpayne opened the pull request:
I made a little and slightly hacky change to support my SSE use-case.
I'm not suggesting that this PR be merged... it's really just here for discussion.

1cg commented on src/htmx.js:
```diff
@@ -1066,6 +1066,10 @@ return (function () {
                     processSSESource(elt, value[1]);
                 }
 
+                if ((value[0] === "forward")) {
```
I like "listen" rather than "forward" here: 

```html 
<div hx-sse="listen"> ... </div> 
```

wtpayne replied:
I don't have a strong opinion on the name, so I defer to you on this. I've made the change as requested.

wtpayne commented:
Record of the conversation from Discord:-
[3:19 PM] wtpayne: Let me give the rationale.
[3:19 PM] wtpayne: So ... I've got a dashboard builder.
[3:19 PM] wtpayne: The dashboard components themselves can be dynamically added, removed and modified.
[3:19 PM] wtpayne: For this, sse:swap is perfect.
[3:20 PM] wtpayne: But I also want the dashboard components to display oscilloscope-like graphs...
[3:20 PM] wtpayne: And for the data for the graph to come over as JSON, so I can use D3.js to do the plotting.
[3:21 PM] wtpayne: So I have some situations where I want to do things the HTMX way ... and have a bit of HTML swapping in.
[3:21 PM] wtpayne: And other situations where I want to do things differently, and have JSON coming in over SSE and being used to update a chart or other visualization with D3.js...
[3:22 PM] wtpayne: So I need to be able to say 'do something with the content of this SSE event ... but don't swap it into the DOM'

1cg commented on src/htmx.js:
```diff
@@ -1081,6 +1085,24 @@ return (function () {
             getInternalData(elt).sseEventSource = source;
         }
 
+        function processSSEForward(elt, sseEventName) {
```
Rather than creating a new method, what about renaming this to something like `processSSEClient` and passing in a swap boolean?  That would reduce some of the code duplication and is about as clear IMO.

wtpayne replied:
Yup. I agree with that as well.

wtpayne replied:
I've made the change.

1cg commented on src/htmx.js:
```diff
@@ -1081,6 +1085,24 @@ return (function () {
+                        sseEventSource.removeEventListener(sseEventName, sseListener);
+                        return;
+                    }
+                    triggerEvent(elt, sseEventName, event)
```
Could we reuse the same code in processSSESwap for triggering the event?

```js
                    triggerEvent(elt, "htmx:sseMessage", event)
```

You would need to look into the `event.detail` object to see the event name, which I guess is a little annoying.

Alternatively, we could adopt this approach (using the SSE event name for the DOM event) in place of the current `htmx:sseMessage` approach in `processSSESwap()`

I don't have strong opinions on this.

wtpayne replied:
I think using htmx:sseMessage minimises the chances of a name collision.

wtpayne replied:
I've made the change.

1cg reviewed (changes requested):
please retarget to `dev` branch as well, thank you for working on this!

wtpayne commented:
> please retarget to `dev` branch as well, thank you for working on this!

Retargeted to 'dev' branch as requested.

wtpayne commented:
Added tests, and checked that tests pass.

wtpayne commented on src/htmx.js:
```diff
@@ -1091,19 +1095,19 @@ return (function () {
+                    }
 
-                    selectAndSwap(swapSpec.swapStyle, elt, target, response, settleInfo)
                     triggerEvent(elt, "htmx:sseMessage", event)
```
I actually think we need a different name for each event, so I'm changing my mind on this. Going to try some experiments first.

1cg replied:
```js
                    triggerEvent(elt, "htmx:sse:" + event.name, event)
```
?

wtpayne replied:
I'm trying something like that ... but I need to do some more investigation first.

benpate replied:
I'm jumping in late here, but wouldn't this make it more difficult to listen to different kinds of sse messages?  If there's just one event name, then I'd only need one function, which could just switch on the message name to differentiate.  Maybe it's an issue of architecture.

1cg commented:
What's the status of this pull request?  Is it ready for review?

wtpayne commented:
I think so. I've been using it like this without too many issue so far.

On Sat, 2 Jan 2021, 00:05 1cg, <notifications@github.com> wrote:

> What's the status of this pull request? Is it ready for review?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/bigskysoftware/htmx/pull/263#issuecomment-753406999>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ53A3KDWRPZWMQWM7URSDSXZPLXANCNFSM4UYIFLDA>
> .
>

wtpayne commented:
However... I did make a fix for what appeared to me to be a bug ... Settle
wasn't being triggered, so I added it in. It seems to work, but I'm not
sure 100% if it's the right thing to do.

On Sat, 2 Jan 2021, 02:15 William Payne, <wtpayne@gmail.com> wrote:

> I think so. I've been using it like this without too many issue so far.
>
> On Sat, 2 Jan 2021, 00:05 1cg, <notifications@github.com> wrote:
>
>> What's the status of this pull request? Is it ready for review?
>>
>> —
>> You are receiving this because you modified the open/close state.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/bigskysoftware/htmx/pull/263#issuecomment-753406999>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAJ53A3KDWRPZWMQWM7URSDSXZPLXANCNFSM4UYIFLDA>
>> .
>>
>

wtpayne commented:
@benpate - I have a separate callback for each type of SSE event, and it would be exceedingly difficult to merge them together. (The callbacks are themselves transmitted over SSE) so my very strong preference would be to leave the SSE event name in the DOM event name if at all possible.

wtpayne reviewed (commented):
I think the call to settleImmediately is the right thing to do ... but I'm not sure.

1cg commented:
resurrecting this thread: @benpate just pulled sse out to a separate extension and I'd like to revist this change in that context

alexpetros commented:
It looks like this is stalled out. It in the time since, Server-Side Events have been pulled out to an extension. If the current SSE extension still has the issue, feel free to either re-open this PR or make a new one and we'll get it looked at ASAP.
